txs/cache.go

```diff
@@ -0,0 +1,671 @@
+			log.Uint64("tx_nonce", ntx.nonce))
+		ac.hasGapInDB = true
+		return errNonceTooBig
+	}
+
+	if ntx.nonce < next {
+		return ac.addToExistingNonce(logger, tp, ntx)
+	}
+
+	// tx uses the next nonce
+	if ac.availBalance < ntx.maxSpending() {
+		// TODO: do better here
+		// this is too harsh. conservative balance does not factor into incoming fund.
+		// a transaction can still be valid.
+		if err := ac.discard(logger, tp, []*nanoTX{ntx}, nil); err != nil {
+			logger.With().Error("failed to discard insufficient balance tx", tx.ID(), log.Err(err))
+			return err
+		}
+		return errInsufficientBalance
+	}
+
+	if err := ac.accept(ntx, ac.availBalance); err != nil {
+		return err
+	}
+
+	if !ac.hasGapInDB {
+		return nil
+	}
+
+	// check DB for more pending nonceTXs with higher nonce
+	byNonce, err := ac.getPendingByNonce(logger, tp)
+	if err != nil {
+		return err
+	}
+	if err = ac.addBatch(logger, tp, byNonce); err != nil {
+		return err
+	}
+	ac.hasGapInDB = false
+	return nil
+}
+
+func (ac *accountCache) getPendingByNonce(logger log.Log, tp txProvider) (map[uint64][]*nanoTX, error) {
+	mtxs, err := tp.getAcctPending(ac.addr)
+	if err != nil {
+		logger.With().Error("failed to get more pending nonceTXs from db", ac.addr, log.Err(err))
+		return nil, err
+	}
+	byNonce := make(map[uint64][]*nanoTX)
+	for _, mtx := range mtxs {
+		if _, ok := byNonce[mtx.AccountNonce]; !ok {
+			byNonce[mtx.AccountNonce] = make([]*nanoTX, 0, maxTXsPerAcct)
+		}
+		byNonce[mtx.AccountNonce] = append(byNonce[mtx.AccountNonce], newNanoTX(mtx))
+	}
+	return byNonce, nil
+}
+
+// find the first nonce without a layer.
+// a nonce with a valid layer indicates that it's already packed in a proposal/block.
+func (ac *accountCache) findMempoolOffset() int {
+	for i, nonceTXs := range ac.txsByNonce {
+		if nonceTXs.layer() == (types.LayerID{}) {
+			return i
+		}
+	}
+	return -1
+}
+
+func (ac *accountCache) getMempool(logger log.Log) ([]*nanoTX, error) {
+	bests := make([]*nanoTX, 0, maxTXsPerAcct)
+	offset := ac.findMempoolOffset()
+	if offset < 0 {
+		return nil, nil
+	}
+	for _, nonceTXs := range ac.txsByNonce[offset:] {
+		if nonceTXs.layer() != (types.LayerID{}) {
+			logger.With().Error("some proposals/blocks packed nonceTXs out of nonce order",
+				nonceTXs.best.tid,
+				nonceTXs.best.layer,
+				nonceTXs.best.block,
+				log.Uint64("nonce", nonceTXs.best.nonce))
+		}
+		bests = append(bests, nonceTXs.best)
+	}
+	return bests, nil
+}
+
+// TODO: check that applyLayer is called in order
+// add last applied layer to SVM?
+func (ac *accountCache) applyLayer(
+	logger log.Log,
+	newNonce, newBalance uint64,
+	tp txProvider,
+	lid types.LayerID,
+	bid types.BlockID,
+	appliedByNonce map[uint64]*nanoTX,
+) error {
+	nextNonce := ac.startNonce
+	var toApply, toDiscard []types.TransactionID
+	for len(appliedByNonce) > 0 {
+		offset := nextNonce - ac.startNonce
+		if _, ok := appliedByNonce[nextNonce]; !ok {
+			allNonce := make([]uint64, 0, len(appliedByNonce))
```

Comment:
allocation should be moved to nonceMarshaller, so at least it will be lazily allocated only when needed

Reply:
done. tho i wish we have `any` available here